In [3]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

### Exercise 1.1

Read in all data sets and inspect their basic properties.


In [4]:
ATHLETES_PATH = "../data/athletes.csv"
GAMES_PATH = "../data/games.csv"
RESULTS_PATH = "../data/results.xlsx"

In [84]:
athletes_df = pd.read_csv(ATHLETES_PATH)
games_df = pd.read_csv(GAMES_PATH,sep=";")

xls = pd.ExcelFile(RESULTS_PATH)


In [6]:
xls.sheet_names

['Year 2002', 'Year 2004', 'Year 2006', 'Year 2008', 'Year 2010', 'Year 2012']

In [7]:
results_dfs = {}
results_dataframe = pd.DataFrame()


for sheet_name in xls.sheet_names:
    results_dfs[sheet_name] = pd.read_excel(RESULTS_PATH, sheet_name=sheet_name)
    print(sheet_name)
    results_dataframe = pd.concat([results_dataframe, results_dfs[sheet_name]], axis=0, ignore_index=True)
    results_dataframe.shape

print(results_dataframe.head())

Year 2002
Year 2004
Year 2006
Year 2008
Year 2010
Year 2012
   Unnamed: 0 Unnamed: 1     Unnamed: 2           Unnamed: 3 Unnamed: 4  \
0  athlete_id     season          sport                event   position   
1       13065     winter      Bobsleigh           Two, Women          7   
2       13139     winter  Speed Skating  3,000 metres, Women         10   
3       13139     winter  Speed Skating  5,000 metres, Women          3   
4       68826     winter      Bobsleigh             Two, Men         10   

  Unnamed: 5     Unnamed: 6    Unnamed: 7            Unnamed: 8  
0      medal        country  country_code                  team  
1        NaN          Italy           ITA  Gerda Weissensteiner  
2        NaN         Canada           CAN                   NaN  
3     Bronze         Canada           CAN                   NaN  
4        NaN  Great Britain           GBR          Lee Johnston  


In [8]:
print(results_dataframe.shape)
results_dataframe.columns = results_dataframe.iloc[0]
results_dataframe = results_dataframe[1:]
results_dataframe.head()

(52333, 9)


0 athlete_id  season          sport                event position   medal  \
1      13065  winter      Bobsleigh           Two, Women        7     NaN   
2      13139  winter  Speed Skating  3,000 metres, Women       10     NaN   
3      13139  winter  Speed Skating  5,000 metres, Women        3  Bronze   
4      68826  winter      Bobsleigh             Two, Men       10     NaN   
5      69119  winter      Bobsleigh            Four, Men       14     NaN   

0        country country_code                  team  
1          Italy          ITA  Gerda Weissensteiner  
2         Canada          CAN                   NaN  
3         Canada          CAN                   NaN  
4  Great Britain          GBR          Lee Johnston  
5  Great Britain          GBR       Great Britain 2

In [92]:
results_dataframe.head()

0 athlete_id  season          sport                event position   medal  \
1      13065  winter      Bobsleigh           Two, Women        7     NaN   
2      13139  winter  Speed Skating  3,000 metres, Women       10     NaN   
3      13139  winter  Speed Skating  5,000 metres, Women        3  Bronze   
4      68826  winter      Bobsleigh             Two, Men       10     NaN   
5      69119  winter      Bobsleigh            Four, Men       14     NaN   

0        country country_code                  team  
1          Italy          ITA  Gerda Weissensteiner  
2         Canada          CAN                   NaN  
3         Canada          CAN                   NaN  
4  Great Britain          GBR          Lee Johnston  
5  Great Britain          GBR       Great Britain 2

In [91]:
results_dataframe['athlete_id'] = results_dataframe['athlete_id'].astype(str)
results_dataframe['athlete_id'] = results_dataframe['athlete_id'].astype(int)

ValueError: invalid literal for int() with base 10: 'athlete_id'

In [9]:
games_df.shape

(62, 6)

In [10]:
games_df.head()

year  season       city        country      opened      closed
0  1912  summer  Stockholm         Sweden  1912-07-06  1912-07-27
1  1920  summer  Antwerpen        Belgium  1920-08-14  1920-08-30
2  1996  summer    Atlanta  United States  1996-07-19  1996-08-04
3  1924  summer      Paris         France  1924-07-05  1924-07-27
4  1992  summer  Barcelona          Spain  1992-07-25  1992-08-09

In [11]:
athletes_df.head()

athlete_id                  name gender    measurements        born died  \
0         7.0         Patrick•Chila   male  180 cm / 73 kg  1969-11-27  NaN   
1        15.0           Damien•Éloi   male  165 cm / 58 kg  1969-07-04  NaN   
2        18.0  Jean-Philippe•Gatien   male  178 cm / 73 kg  1968-10-16  NaN   
3        27.0     Christophe•Legoût   male  177 cm / 75 kg  1973-08-06  NaN   
4        35.0       Fabrice•Santoro   male  177 cm / 68 kg  1972-12-09  NaN   

                                        affiliations  
0  Espérance de Reuilly / Levallois SCTT / Royal ...  
1  Caen TTC, Caen (FRA) / SAG Cestas, Cestas (FRA...  
2               Levallois SC, Levallois-Perret (FRA)  
3  Levallois SC, Levallois-Perret (FRA) / Istres ...  
4                                                NaN

In [80]:
athletes_df.shape

(39375, 7)

In [12]:
athletes_df.isna().sum()


athlete_id          0
name                0
gender              0
measurements      378
born                2
died            39073
affiliations     9179
dtype: int64

In [ ]:
results_dataframe['athlete_id'].isna()

AttributeError: 'Series' object has no attribute 'true'

In [95]:
results_dataframe['athlete_id'] = pd.to_numeric(results_dataframe['athlete_id'], errors='coerce')
results_dataframe['athlete_id'] = results_dataframe['athlete_id'].astype(int)

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

In [94]:
results_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52332 entries, 1 to 52332
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   athlete_id    52327 non-null  float64
 1   season        52332 non-null  object 
 2   sport         52332 non-null  object 
 3   event         52332 non-null  object 
 4   position      43161 non-null  object 
 5   medal         7271 non-null   object 
 6   country       52332 non-null  object 
 7   country_code  52332 non-null  object 
 8   team          20087 non-null  object 
dtypes: float64(1), object(8)
memory usage: 3.6+ MB


In [ ]:
athletes_df['athlete_id'] = athletes_df['athlete_id'].astype(int)
athletes_df['athlete_id']

0             7
1            15
2            18
3            27
4            35
          ...  
39370    138425
39371    138426
39372    138427
39373    138428
39374    138429
Name: athlete_id, Length: 39375, dtype: int64

In [13]:

results_dataframe.isna().sum()

0
athlete_id          0
season              0
sport               0
event               0
position         9171
medal           45061
country             0
country_code        0
team            32245
dtype: int64

In [14]:
games_df.isna().sum()


year        0
season      0
city        0
country     0
opened     11
closed     13
dtype: int64

In [15]:
athletes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39375 entries, 0 to 39374
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   athlete_id    39375 non-null  float64
 1   name          39375 non-null  object 
 2   gender        39375 non-null  object 
 3   measurements  38997 non-null  object 
 4   born          39373 non-null  object 
 5   died          302 non-null    object 
 6   affiliations  30196 non-null  object 
dtypes: float64(1), object(6)
memory usage: 2.1+ MB


In [16]:
games_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62 entries, 0 to 61
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   year     62 non-null     int64 
 1   season   62 non-null     object
 2   city     62 non-null     object
 3   country  62 non-null     object
 4   opened   51 non-null     object
 5   closed   49 non-null     object
dtypes: int64(1), object(5)
memory usage: 3.0+ KB


In [17]:
results_dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52332 entries, 1 to 52332
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   athlete_id    52332 non-null  object
 1   season        52332 non-null  object
 2   sport         52332 non-null  object
 3   event         52332 non-null  object
 4   position      43161 non-null  object
 5   medal         7271 non-null   object
 6   country       52332 non-null  object
 7   country_code  52332 non-null  object
 8   team          20087 non-null  object
dtypes: object(9)
memory usage: 3.6+ MB


In [18]:
results_dataframe.describe()

0       athlete_id  season      sport            event  position   medal  \
count        52332   52332      52332            52332     43161    7271   
unique       29339       3         57              352       209       4   
top          93860  summer  Athletics  Ice Hockey, Men         5  Bronze   
freq            23   39478       6774              868      2230    2507   

0             country country_code           team  
count           52332        52332          20087  
unique            207          207           3995  
top     United States          USA  United States  
freq             3176         3176           1089

### Exercise 1.2

**Combine all provided data suitably into a single, tidy data set (e.g. by merging, concatenating or reshaping the data) such that you are able to answer the subsequent questions.**


In [19]:
results_athlethes = results_dataframe.merge(athletes_df, how="left", left_on="athlete_id", right_on="athlete_id")

In [20]:
results_athlethes.head()

athlete_id  season          sport                event position   medal  \
0      13065  winter      Bobsleigh           Two, Women        7     NaN   
1      13139  winter  Speed Skating  3,000 metres, Women       10     NaN   
2      13139  winter  Speed Skating  5,000 metres, Women        3  Bronze   
3      68826  winter      Bobsleigh             Two, Men       10     NaN   
4      69119  winter      Bobsleigh            Four, Men       14     NaN   

         country country_code                  team                name  \
0          Italy          ITA  Gerda Weissensteiner  Antonella•Bellutti   
1         Canada          CAN                   NaN        Clara•Hughes   
2         Canada          CAN                   NaN        Clara•Hughes   
3  Great Britain          GBR          Lee Johnston         Marcus•Adam   
4  Great Britain          GBR       Great Britain 2        Phil•Harries   

   gender    measurements        born died  \
0  female  180 cm / 73 kg  1968-11-07  NaN   
1  female  175 cm / 72 kg  1972-09-27  NaN   
2  female  175 cm / 72 kg  1972-09-27  NaN   
3    male  182 cm / 82 kg  1968-02-28  NaN   
4    male  183 cm / 88 kg  1966-04-07  NaN   

                                        affiliations  
0                    S.C. Adriana Bolzano/BC Cortina  
1  Manitoba / Saturn Racing Team / Sherbrooke Spe...  
2  Manitoba / Saturn Racing Team / Sherbrooke Spe...  
3                 Belgrave Harriers, Wimbledon (GBR)  
4                     Derby & County AC, Derby (GBR)

In [21]:
results_athlethes.drop_duplicates(inplace=True)
results_athlethes.shape

(50681, 15)

In [22]:
results_with_games = results_athlethes.merge(games_df, how="left", on=["season","country"])

In [23]:
results_with_games.head()

athlete_id  season          sport                event position medal  \
0      13065  winter      Bobsleigh           Two, Women        7   NaN   
1      13065  winter      Bobsleigh           Two, Women        7   NaN   
2      13065  winter      Bobsleigh           Two, Women        7   NaN   
3      13065  winter      Bobsleigh           Two, Women        7   NaN   
4      13139  winter  Speed Skating  3,000 metres, Women       10   NaN   

  country country_code                  team                name  gender  \
0   Italy          ITA  Gerda Weissensteiner  Antonella•Bellutti  female   
1   Italy          ITA  Gerda Weissensteiner  Antonella•Bellutti  female   
2   Italy          ITA  Gerda Weissensteiner  Antonella•Bellutti  female   
3   Italy          ITA  Gerda Weissensteiner  Antonella•Bellutti  female   
4  Canada          CAN                   NaN        Clara•Hughes  female   

     measurements        born died  \
0  180 cm / 73 kg  1968-11-07  NaN   
1  180 cm / 73 kg  1968-11-07  NaN   
2  180 cm / 73 kg  1968-11-07  NaN   
3  180 cm / 73 kg  1968-11-07  NaN   
4  175 cm / 72 kg  1972-09-27  NaN   

                                        affiliations    year  \
0                    S.C. Adriana Bolzano/BC Cortina  2006.0   
1                    S.C. Adriana Bolzano/BC Cortina  1956.0   
2                    S.C. Adriana Bolzano/BC Cortina  1944.0   
3                    S.C. Adriana Bolzano/BC Cortina  2026.0   
4  Manitoba / Saturn Racing Team / Sherbrooke Spe...  1988.0   

                       city      opened      closed  
0                    Torino  2006-02-10  2006-02-26  
1         Cortina d'Ampezzo  1956-01-26  1956-02-05  
2         Cortina d'Ampezzo         NaN         NaN  
3  Milano-Cortina d'Ampezzo         NaN         NaN  
4                   Calgary  1988-02-13  1988-02-28

In [24]:
results_with_games.drop_duplicates(inplace=True)
results_with_games.isna().sum()
# results_with_games.shape


athlete_id          0
season              0
sport               0
event               0
position        12322
medal           67965
country             0
country_code        0
team            48220
name               52
gender             52
measurements      350
born               54
died            81453
affiliations    11651
year            24965
city            24965
opened          37153
closed          40718
dtype: int64

In [25]:
results_with_games.shape

(81994, 19)

### Exercise 1.3

Describe and justify your data processing steps in a few sentences: How do you combine the data? What are the main challenges? How do you deal with these challenges and why?


**Data Processing Methodology**

1. First, I merged the results_dataframe with athletes_df using a left join on athlete_id
   This preserves all Olympic results while adding athlete details where available

2. Then merged the resulting dataframe with games_df using a left join on season and country
   This adds Olympic Games information like year, city, opening/closing dates

**Main challenges and solutions:**

1. Data quality issues: Many NA values in several columns (visible in isna().sum() outputs)
   Solution: Used left joins to preserve all result records even if athlete or games info missing
2. Data integration: Results data came from multiple Excel sheets
   Solution: Used pd.concat to combine all sheets into single results_dataframe
3. Different data formats: Excel sheets had headers as first row rather than column names
   Solution: Used iloc[0] to set column names and removed that row from the data
4. Inconsistent keys between dataframes: Had to identify common fields for merging
   Solution: Used season and country as join keys for games data as there was no direct ID field
   The final dataset contains complete Olympic results with athlete and games information


### Exercise 1.4

Sort the data permanently by year, sport, event and position. Then display the first 3 rows of the data set, and print the number of rows and columns.


In [26]:
results_with_games.sort_values(by=['year','sport','event','position'], inplace=True)

In [27]:
results_with_games.head(3)

athlete_id  season    sport            event position medal country  \
15473     102871  summer  Archery  Individual, Men       29   NaN  Greece   
15469     102870  summer  Archery  Individual, Men       54   NaN  Greece   
15477     102872  summer  Archery  Individual, Men       57   NaN  Greece   

      country_code team                     name gender    measurements  \
15473          GRE  NaN  Alexandros•Karageorgiou   male  172 cm / 58 kg   
15469          GRE  NaN   Georgios•Kalogiannidis   male  180 cm / 73 kg   
15477          GRE  NaN          Apostolos•Nanos   male  183 cm / 85 kg   

             born died                                       affiliations  \
15473  1986-06-03  NaN  Aristotle University Alumni Archery Club, Thes...   
15469  1982-11-21  NaN  Aristotle University Alumni Archery Club, Thes...   
15477  1966-02-05  NaN                                                NaN   

         year    city      opened      closed  
15473  1896.0  Athina  1896-04-06  1896-04-15  
15469  1896.0  Athina  1896-04-06  1896-04-15  
15477  1896.0  Athina  1896-04-06  1896-04-15

In [28]:
results_with_games.shape

(81994, 19)

# Exercise 2


### Exercise 2.1

**Which lastname is the most common among the athletes of the 2012 olympic games, and how often does this name occur?**


In [29]:
df = results_with_games.copy()

In [30]:
df.head()

athlete_id  season    sport              event position medal country  \
15473     102871  summer  Archery    Individual, Men       29   NaN  Greece   
15469     102870  summer  Archery    Individual, Men       54   NaN  Greece   
15477     102872  summer  Archery    Individual, Men       57   NaN  Greece   
10176      80496  summer  Archery  Individual, Women        7   NaN  Greece   
38189     102837  summer  Archery  Individual, Women       11   NaN  Greece   

      country_code team                     name  gender    measurements  \
15473          GRE  NaN  Alexandros•Karageorgiou    male  172 cm / 58 kg   
15469          GRE  NaN   Georgios•Kalogiannidis    male  180 cm / 73 kg   
15477          GRE  NaN          Apostolos•Nanos    male  183 cm / 85 kg   
10176          GRE  NaN         Evangelia•Psarra  female  172 cm / 63 kg   
38189          GRE  NaN          Elpida•Romantzi  female  167 cm / 53 kg   

             born died                                       affiliations  \
15473  1986-06-03  NaN  Aristotle University Alumni Archery Club, Thes...   
15469  1982-11-21  NaN  Aristotle University Alumni Archery Club, Thes...   
15477  1966-02-05  NaN                                                NaN   
10176  1974-06-17  NaN  Aristtotle University Alumni Archery Club, The...   
38189  1981-07-17  NaN  Aristotle University Alumni Archery Club, Thes...   

         year    city      opened      closed  
15473  1896.0  Athina  1896-04-06  1896-04-15  
15469  1896.0  Athina  1896-04-06  1896-04-15  
15477  1896.0  Athina  1896-04-06  1896-04-15  
10176  1896.0  Athina  1896-04-06  1896-04-15  
38189  1896.0  Athina  1896-04-06  1896-04-15

In [31]:
df['Name'] = df['name'].str.split('•',n=1, expand=True)[0]
df['Surname'] = df['name'].str.split('•',n=1, expand=True)[1]
df.drop(columns=['name'], inplace=True)

In [32]:
df_2012 = df[df['year'] == 2012]
df_2012['Surname'].value_counts().head(10)


Surname
Smith       26
Thomas      12
Halsall     11
Murray      11
Jackson     11
Williams    10
Tweddle     10
Clark       10
Turner       8
Cook         8
Name: count, dtype: int64

**Answer: Most Common Surname for olympic games of 2012 is Smith and it occurs 26 times.**

**My approuch**

- I splitted name column to retrieve surname
- then I filter by date
- and finally count values to see number of their appearence


### Exercise 2.2

**Which woman won the Gold medal in the 100 meters race of the 2012 olympic games?**


In [33]:
df_2012['event']

15463                  Individual, Men
42477                  Individual, Men
38202                  Individual, Men
35317                  Individual, Men
62082                  Individual, Men
                     ...              
72973          Middle-Heavyweight, Men
72977                Middleweight, Men
38757              Middleweight, Women
81228    Lightweight, Freestyle, Women
25695     Middleweight, Freestyle, Men
Name: event, Length: 1444, dtype: object

In [42]:
print(df_2012['event'].unique())
len(df_2012['event'].unique())

['Individual, Men' 'Individual, Women' 'Team, Men' 'Team, Women'
 'Balance Beam, Women' 'Floor Exercise, Men' 'Floor Exercise, Women'
 'Horizontal Bar, Men' 'Horse Vault, Men' 'Horse Vault, Women'
 'Individual All-Around, Men' 'Individual All-Around, Women'
 'Parallel Bars, Men' 'Pommelled Horse, Men' 'Rings, Men'
 'Team All-Around, Men' 'Team All-Around, Women' 'Uneven Bars, Women'
 'Duet, Women' '1,500 metres, Men' '1,500 metres, Women'
 '10,000 metres, Men' '10,000 metres, Women' '100 metres Hurdles, Women'
 '100 metres, Men' '100 metres, Women' '110 metres Hurdles, Men'
 '20 kilometres Walk, Women' '200 metres, Men' '200 metres, Women'
 '3,000 metres Steeplechase, Men' '3,000 metres Steeplechase, Women'
 '4 × 100 metres Relay, Men' '4 × 100 metres Relay, Women'
 '4 × 400 metres Relay, Men' '4 × 400 metres Relay, Women'
 '400 metres Hurdles, Men' '400 metres Hurdles, Women' '400 metres, Men'
 '400 metres, Women' '5,000 metres, Men' '5,000 metres, Women'
 '50 kilometres Walk, Men' '8

242

In [41]:
df_2012['event']

15463                  Individual, Men
42477                  Individual, Men
38202                  Individual, Men
35317                  Individual, Men
62082                  Individual, Men
                     ...              
72973          Middle-Heavyweight, Men
72977                Middleweight, Men
38757              Middleweight, Women
81228    Lightweight, Freestyle, Women
25695     Middleweight, Freestyle, Men
Name: event, Length: 1444, dtype: object

In [44]:
df_2012[(df_2012['event'] == '100 metres, Women')]

athlete_id  season      sport              event position medal  \
48251     114716  summer  Athletics  100 metres, Women        6   NaN   
19219     104922  summer  Athletics  100 metres, Women      NaN   NaN   
48291     114742  summer  Athletics  100 metres, Women      NaN   NaN   
48299     114747  summer  Athletics  100 metres, Women      NaN   NaN   
81287     126945  summer  Athletics  100 metres, Women      NaN   NaN   

             country country_code team  gender    measurements        born  \
48251  Great Britain          GBR  NaN  female  163 cm / 60 kg  1983-03-20   
19219  Great Britain          GBR  NaN  female  165 cm / 55 kg  1979-12-30   
48291  Great Britain          GBR  NaN  female  168 cm / 57 kg  1982-08-12   
48299  Great Britain          GBR  NaN  female  175 cm / 67 kg  1986-01-24   
81287  Great Britain          GBR  NaN  female  178 cm / 76 kg  1984-10-28   

      died                                      affiliations    year    city  \
48251  NaN  Woodford Green with Essex Ladies, Woodford (GBR)  2012.0  London   
19219  NaN         Shaftesbury Barnet Harriers, London (GBR)  2012.0  London   
48291  NaN                     Harrow AC, Harrow Weald (GBR)  2012.0  London   
48299  NaN     Blackheath & Bromley Harriers AC, Hayes (GBR)  2012.0  London   
81287  NaN               Liverpool Harriers, Liverpool (GBR)  2012.0  London   

           opened      closed       Name   Surname  
48251  2012-07-27  2012-08-12  Jeannette    Kwakye  
19219  2012-07-27  2012-08-12        Abi  Oyepitan  
48291  2012-07-27  2012-08-12      Laura    Turner  
48299  2012-07-27  2012-08-12    Montell   Douglas  
81287  2012-07-27  2012-08-12     Anyika    Onuora

In [45]:
df_2012[(df_2012['event'] == '100 metres, Women') & (df_2012['medal']=='Gold')]

Empty DataFrame
Columns: [athlete_id, season, sport, event, position, medal, country, country_code, team, gender, measurements, born, died, affiliations, year, city, opened, closed, Name, Surname]
Index: []

**Answer: Nobody got Gold Medal**

**My approuch**

- I checken event column and it unique values to see how data is saved its 242 unique value here
- then I try to filter out our required data point but I have found there is 6 cases of '100 metres, Women' but
- there is no gold medal winner in '100 metres, Women' :(


### Exercise 2.3

**Which athlete has won the largest number of gold medals (in all years covered by the dataset), considering only athletes from the following countries: “Jamaica”, “Trinidad and Tobago”, “Bar-bados”, “Grenada”, “Saint Kitts and Nevis”?**


In [36]:
df['country'].unique()

array(['Greece', 'France', 'United States', 'Great Britain', 'Sweden',
       'Germany', 'Belgium', 'Switzerland', 'Finland', 'Italy', 'Norway',
       'Australia', 'Austria', 'Japan', 'Mexico', 'Canada',
       'Republic of Korea', 'Spain', "People's Republic of China",
       'Russian Federation', 'Brazil', 'Croatia', 'Czech Republic',
       'Slovenia', 'Bulgaria', 'Argentina', 'Slovakia', 'Ukraine',
       'Andorra', 'Liechtenstein', 'Kazakhstan', 'Romania', 'Ireland',
       'South Africa', 'Bosnia and Herzegovina', 'Latvia', 'Iceland',
       'Poland', 'Chile', 'Monaco', 'Hungary', 'Lebanon',
       'Serbia and Montenegro', 'Tajikistan', 'Denmark', 'Algeria',
       'Georgia', 'Israel', 'Cyprus', 'Albania',
       'Islamic Republic of Iran', 'Madagascar', 'Kyrgyzstan',
       'Republic of Moldova', 'New Zealand', 'North Macedonia', 'Fiji',
       'San Marino', 'Lithuania', 'Montenegro', 'Estonia', 'Peru',
       'Cayman Islands', 'Azerbaijan', 'Senegal', 'Morocco', 'Uzbekistan',


In [46]:
l = ['Jamaica', 'Trinidad and Tobago', 'Barbados', 'Grenada', 'Saint Kitts and Nevis']
for c in l:
    if c in df['country'].unique():
        print(c)
    else:
        print(f"{c} not in df_2012['country'].unique()")

""" 
checking because barbados was written as bar-bados in the dataset 
"""

Jamaica
Trinidad and Tobago
Barbados
Grenada
Saint Kitts and Nevis


' \nchecking because barbados was written as bar-bados in the dataset \n'

In [49]:
df_gold = df[ (df['country'].isin(l)) & (df['medal'] == 'Gold') ]
df_gold.groupby('athlete_id')['medal'].count().sort_values(ascending=False)

athlete_id
105512    5
114740    2
87116     2
87125     1
72224     1
89036     1
104889    1
105495    1
104904    1
114541    1
114925    1
115065    1
115511    1
115522    1
115566    1
120717    1
120724    1
125971    1
126065    1
Name: medal, dtype: int64

In [51]:
df.loc[df['athlete_id'] == 105512] 

athlete_id  season      sport                      event position medal  \
39953     105512  summer  Athletics            100 metres, Men        1  Gold   
64346     105512  summer  Athletics            100 metres, Men        1  Gold   
39954     105512  summer  Athletics            200 metres, Men        1  Gold   
64347     105512  summer  Athletics            200 metres, Men        1  Gold   
20273     105512  summer  Athletics            200 metres, Men      NaN   NaN   
64348     105512  summer  Athletics  4 × 100 metres Relay, Men        1  Gold   
39955     105512  summer  Athletics  4 × 100 metres Relay, Men      NaN   NaN   

       country country_code     team gender    measurements        born died  \
39953  Jamaica          JAM      NaN   male  196 cm / 95 kg  1986-08-21  NaN   
64346  Jamaica          JAM      NaN   male  196 cm / 95 kg  1986-08-21  NaN   
39954  Jamaica          JAM      NaN   male  196 cm / 95 kg  1986-08-21  NaN   
64347  Jamaica          JAM      NaN   male  196 cm / 95 kg  1986-08-21  NaN   
20273  Jamaica          JAM      NaN   male  196 cm / 95 kg  1986-08-21  NaN   
64348  Jamaica          JAM  Jamaica   male  196 cm / 95 kg  1986-08-21  NaN   
39955  Jamaica          JAM  Jamaica   male  196 cm / 95 kg  1986-08-21  NaN   

                            affiliations  year city opened closed   Name  \
39953  Racers Track Club, Kingston (JAM)   NaN  NaN    NaN    NaN  Usain   
64346  Racers Track Club, Kingston (JAM)   NaN  NaN    NaN    NaN  Usain   
39954  Racers Track Club, Kingston (JAM)   NaN  NaN    NaN    NaN  Usain   
64347  Racers Track Club, Kingston (JAM)   NaN  NaN    NaN    NaN  Usain   
20273  Racers Track Club, Kingston (JAM)   NaN  NaN    NaN    NaN  Usain   
64348  Racers Track Club, Kingston (JAM)   NaN  NaN    NaN    NaN  Usain   
39955  Racers Track Club, Kingston (JAM)   NaN  NaN    NaN    NaN  Usain   

      Surname  
39953    Bolt  
64346    Bolt  
39954    Bolt  
64347    Bolt  
20273    Bolt  
64348    Bolt  
39955    Bolt

**Answer: Nobody got Gold Medal from this countries**
**My approuch**

- As always I checked out column first to understand whats happend inside so
- I filtered by required countries
- and finally it showed that there is athlete with athlete_id:105512 who won maximum amount of 5 gold medals from this specified countries
- and wow this athlete is Usain Bolt


### Exercise 2.4

**What was the best position, worst position, median position achieved by athletes from “Mongolia”?**


In [ ]:
df_temp = df[df['country'] == 'Mongolia']['position']

In [ ]:
df_temp = pd.to_numeric(df_temp, errors='coerce')

best_position = df_temp.min()
worst_position = df_temp.max()
median_position = df_temp.median()

print(f"Best position: {best_position}")
print(f"Worst position: {worst_position}")
print(f"Median position: {median_position}")

74037    33.0
74036     9.0
48460     NaN
41340    52.0
26004    75.0
         ... 
52257    10.0
75377     3.0
25760    12.0
75199     5.0
52323     8.0
Name: position, Length: 92, dtype: float64 
 <class 'pandas.core.series.Series'>
Best position: 1.0
Worst position: 98.0
Median position: 13.0


**Answer: For Mongolia Best position: 1.0 Worst position: 98.0 Median position: 13.0**

**My approuch**

- I filter seperated dataset for MOngolia and position column
- converting to numeric values because values are saved in column in string format
- then comes mathematical exploration using functions
- and ofc printing out values


# Exercise 3


### Exercise 3.1

Which are the 5 sports with the highest average age of gold medal winners? Provide the sport and
the average age. (Note: We define age as the number of completed years between the athlete’s
birth and the opening date of the olympic games.)


In [61]:
df.head()

athlete_id  season    sport              event position medal country  \
15473     102871  summer  Archery    Individual, Men       29   NaN  Greece   
15469     102870  summer  Archery    Individual, Men       54   NaN  Greece   
15477     102872  summer  Archery    Individual, Men       57   NaN  Greece   
10176      80496  summer  Archery  Individual, Women        7   NaN  Greece   
38189     102837  summer  Archery  Individual, Women       11   NaN  Greece   

      country_code team  gender    measurements        born died  \
15473          GRE  NaN    male  172 cm / 58 kg  1986-06-03  NaN   
15469          GRE  NaN    male  180 cm / 73 kg  1982-11-21  NaN   
15477          GRE  NaN    male  183 cm / 85 kg  1966-02-05  NaN   
10176          GRE  NaN  female  172 cm / 63 kg  1974-06-17  NaN   
38189          GRE  NaN  female  167 cm / 53 kg  1981-07-17  NaN   

                                            affiliations    year    city  \
15473  Aristotle University Alumni Archery Club, Thes...  1896.0  Athina   
15469  Aristotle University Alumni Archery Club, Thes...  1896.0  Athina   
15477                                                NaN  1896.0  Athina   
10176  Aristtotle University Alumni Archery Club, The...  1896.0  Athina   
38189  Aristotle University Alumni Archery Club, Thes...  1896.0  Athina   

           opened      closed        Name        Surname  
15473  1896-04-06  1896-04-15  Alexandros   Karageorgiou  
15469  1896-04-06  1896-04-15    Georgios  Kalogiannidis  
15477  1896-04-06  1896-04-15   Apostolos          Nanos  
10176  1896-04-06  1896-04-15   Evangelia         Psarra  
38189  1896-04-06  1896-04-15      Elpida       Romantzi

In [64]:
df_temp_for_ex3 = df.copy()

In [ ]:
df_gold_winners = df_temp_for_ex3[df_temp_for_ex3['medal'] == 'Gold'].copy()
df_gold_winners['opened'] = pd.to_datetime(df_gold_winners['opened'], format='%Y-%m-%d', errors='coerce')
df_gold_winners['born'] = pd.to_datetime(df_gold_winners['born'], format='%Y-%m-%d', errors='coerce')
df_gold_winners.head()


athlete_id  season                sport                          event  \
12375      91346  summer  Artistic Gymnastics                     Rings, Men   
20024     105360  summer            Athletics      20 kilometres Walk, Women   
19354     105013  summer            Athletics      400 metres Hurdles, Women   
8986       47999  summer               Diving  Synchronized Springboard, Men   
11854      90586  summer               Diving  Synchronized Springboard, Men   

      position medal country country_code                team  gender  \
12375        1  Gold  Greece          GRE                 NaN    male   
20024        1  Gold  Greece          GRE                 NaN  female   
19354        1  Gold  Greece          GRE                 NaN  female   
8986         1  Gold  Greece          GRE        Thomas Bimis    male   
11854        1  Gold  Greece          GRE  Nikolaos Siranidis    male   

         measurements       born died                   affiliations    year  \
12375  163 cm / 59 kg 1976-11-12  NaN  Spartakos, Thessaloniki (GRE)  1896.0   
20024  158 cm / 47 kg 1982-01-02  NaN               Asteras Prevezas  1896.0   
19354  174 cm / 60 kg 1979-02-02  NaN              Pelasgos Larissas  1896.0   
8986   175 cm / 75 kg 1976-02-26  NaN  Novolou and Argonautes, Volos  1896.0   
11854  169 cm / 69 kg 1975-06-11  NaN             ANOG, Athina (GRE)  1896.0   

         city     opened      closed         Name     Surname  
12375  Athina 1896-04-06  1896-04-15  Dimosthenis    Tambakos  
20024  Athina 1896-04-06  1896-04-15    Athanasia  Tsoumeleka  
19354  Athina 1896-04-06  1896-04-15         Fani     Khalkia  
8986   Athina 1896-04-06  1896-04-15     Nikolaos   Siranidis  
11854  Athina 1896-04-06  1896-04-15       Thomas       Bimis

In [79]:
df_gold_winners['average_age'] = df_gold_winners['opened'].dt.year - df_gold_winners['born'].dt.year
df_gold_winners['average_age'] = df_gold_winners['average_age'].astype('Int64')
df_gold_winners[df_gold_winners['average_age']>0]

athlete_id  season                 sport  \
1715       99487  winter         Alpine Skiing   
28624      99428  winter         Alpine Skiing   
682        85391  winter         Alpine Skiing   
28626      99428  winter         Alpine Skiing   
2655      100066  winter  Cross Country Skiing   
...          ...     ...                   ...   
24928     107550  summer             Wrestling   
64895     107550  summer             Wrestling   
24948     107562  summer             Wrestling   
64899     107562  summer             Wrestling   
70352     120624  summer             Wrestling   

                                event position medal  country country_code  \
1715                    Downhill, Men        1  Gold  Austria          AUT   
28624                 Downhill, Women        1  Gold  Austria          AUT   
682                 Giant Slalom, Men        1  Gold  Austria          AUT   
28626                  Super G, Women        1  Gold  Austria          AUT   
2655               30 kilometres, Men        1  Gold  Austria          AUT   
...                               ...      ...   ...      ...          ...   
24928   Lightweight, Freestyle, Women        1  Gold    Japan          JPN   
64895   Lightweight, Freestyle, Women        1  Gold    Japan          JPN   
24948  Middleweight, Freestyle, Women        1  Gold    Japan          JPN   
64899  Middleweight, Freestyle, Women        1  Gold    Japan          JPN   
70352    Welterweight, Freestyle, Men        1  Gold    Japan          JPN   

      team  gender  ...       born died  \
1715   NaN    male  ... 1972-08-24  NaN   
28624  NaN  female  ... 1973-03-25  NaN   
682    NaN    male  ... 1969-03-24  NaN   
28626  NaN  female  ... 1973-03-25  NaN   
2655   NaN    male  ... 1974-12-22  NaN   
...    ...     ...  ...        ...  ...   
24928  NaN  female  ... 1982-10-05  NaN   
64895  NaN  female  ... 1982-10-05  NaN   
24948  NaN  female  ... 1984-06-13  NaN   
64899  NaN  female  ... 1984-06-13  NaN   
70352  NaN    male  ... 1986-08-05  NaN   

                                            affiliations    year       city  \
1715                                   Skiclub Steinfeld  1976.0  Innsbruck   
28624               WSV Pernitz-Unterberg, Pernitz (AUT)  1976.0  Innsbruck   
682             WSV Zell am Ziller, Zell am Ziller (AUT)  1976.0  Innsbruck   
28626               WSV Pernitz-Unterberg, Pernitz (AUT)  1976.0  Innsbruck   
2655             SU Böhmerwald, Aigen im Mühlkreis (AUT)  1976.0  Innsbruck   
...                                                  ...     ...        ...   
24928  High School Wrestling Club Hisai/Sogho Securit...  2020.0      Tokyo   
64895  High School Wrestling Club Hisai/Sogho Securit...  2020.0      Tokyo   
24948                  High School Wrestling Club Nagoya  2020.0      Tokyo   
64899                  High School Wrestling Club Nagoya  2020.0      Tokyo   
70352   Self-Defense Forces Sports School , Saitama, JPN  2020.0      Tokyo   

          opened      closed       Name      Surname average_age  
1715  1976-02-04  1976-02-15      Fritz       Strobl           4  
28624 1976-02-04  1976-02-15   Michaela  Dorfmeister           3  
682   1976-02-04  1976-02-15    Stephan   Eberharter           7  
28626 1976-02-04  1976-02-15   Michaela  Dorfmeister           3  
2655  1976-02-04  1976-02-15  Christian     Hoffmann           2  
...          ...         ...        ...          ...         ...  
24928 2021-07-23  2021-08-08      Saori      Yoshida          39  
64895 2021-07-23  2021-08-08      Saori      Yoshida          39  
24948 2021-07-23  2021-08-08      Kaori         Icho          37  
64899 2021-07-23  2021-08-08      Kaori         Icho          37  
70352 2021-07-23  2021-08-08  Tatsuhiro    Yonemitsu          35  

[1347 rows x 21 columns]

In [75]:
df_gold_winners['average_age']

12375   -29439 days
20024   -31316 days
19354   -30251 days
8986    -29179 days
11854   -28919 days
            ...    
41052           NaT
64908           NaT
9479            NaT
52319           NaT
24987           NaT
Name: average_age, Length: 4727, dtype: timedelta64[ns]

**Answer: For Mongolia Best position: 1.0 Worst position: 98.0 Median position: 13.0**

**My approuch**

- I filter seperated dataset for MOngolia and position column
- converting to numeric values because values are saved in column in string format
- then comes mathematical exploration using functions
- and ofc printing out values


### Exercise 3.2

The Body-Mass-Index (BMI) is a rule-of-thumb measure to classify people into groups such as
“underweight”, “normal”, “overweight” etc. The BMI is defined as the body mass (in kilograms)
divided by the square of the body height (in metres), and is expressed in units of kg/m2. Derive
a new column bmi from the provided data. Display the 5 sports with the lowest average BMI of
male athletes.


**Answer: For Mongolia Best position: 1.0 Worst position: 98.0 Median position: 13.0**

**My approuch**

- I filter seperated dataset for MOngolia and position column
- converting to numeric values because values are saved in column in string format
- then comes mathematical exploration using functions
- and ofc printing out values


# Exercise 4


### Exercise 4.1

Calculate the medal table for the olympic games 2012. See the official medal table as a reference. Your
medal table should have the same structure (same column names and same sorting) as the official medal
table. Display the top 10 countries.

Hint: In team events such as Basketball, all team members of the winning team receive a gold medal,
but for the medal table it should only count as one gold medal. The same logic also applies to silver and
bronze medals. You can recognize team events by the fact that multiple athletes of a country have won
the same medal in a single competition.


**Answer: For Mongolia Best position: 1.0 Worst position: 98.0 Median position: 13.0**

**My approuch**

- I filter seperated dataset for MOngolia and position column
- converting to numeric values because values are saved in column in string format
- then comes mathematical exploration using functions
- and ofc printing out values


# Exercise 5


### Exercise 5.1

Identify an interesting pattern, trend, or relationship in the data. This could relate to countries,
sports, gender, time periods, medal counts, or any other aspect. Create a single visualization that
clearly and effectively communicates the insight you found. You are free to choose the type of
chart, but make sure it suits the message you want to convey.


**Answer: For Mongolia Best position: 1.0 Worst position: 98.0 Median position: 13.0**

**My approuch**

- I filter seperated dataset for MOngolia and position column
- converting to numeric values because values are saved in column in string format
- then comes mathematical exploration using functions
- and ofc printing out values


### Exercise 5.2

Briefly describe what the visualization shows and explain why the pattern is interesting/surprising.


**Answer: For Mongolia Best position: 1.0 Worst position: 98.0 Median position: 13.0**

**My approuch**

- I filter seperated dataset for MOngolia and position column
- converting to numeric values because values are saved in column in string format
- then comes mathematical exploration using functions
- and ofc printing out values
